### DISCUSSION TITLE
Many to many to the same table

### DISCUSSION THREAD

**Discussion by hazae41** [supabase/supabase#1969] [Questions] [answered]
I am trying to select rows from a many-to-many relationship

I have three tables 
- "posts" which contains a "author" (fkey to users) and a "text"
- "users" which contains a "name" and "avatar"
- "subscriptions" where each row is a subscription from an user "subscriber" (fkey to users) to another user "subscribed" (fkey to users)

How can I get all posts made by users I subscribed to? (and getting the author's name and avatar for each post)

Thanks a lot 🙏

**Comment by steve-chavez** [comment, marked as answer]
So there is a many-to-many relationship from `users` to `users`(same table) by using `subscriptions` as a join table right?

What you could do right now is to go through the `subscriptions` table explicitly:

```js
const { data, error } = await supabase
  .from('users')
  .select(`
    subscriptions!subscriber(
      subscribed(*,posts(*))
    )
  `)
  .eq('id', 2) // 2 would be your id

//the output should be something like

[
  {
    "subscriptions": [
      {
        "subscribed": {
          "id": 1,
          "name": "user-1",
          "avatar": null,
          "posts": [
            {
              "id": 1,
              "content": "post-1-1",
              "author": 1
            },
            {
              "id": 2,
              "content": "post-2-1",
              "author": 1
            }
          ]
        }
      }
    ]
  }
]
```

**Reply by hazae41** [reply]
Thanks for the reply

> So there is a many-to-many relationship from users to users(same table) by using subscriptions as a join table right?

Yes

However, I would expect something like a list of posts instead of a list of subscriptions

GET /posts/subscribed/myUserName

```JSON
[{
  "text": "blabla",
  "author": {
     "name": "someUser",
     "avatar": "someUser.png"
   }
},
{
  "text": "hello world",
  "author": {
     "name": "someOtherUser",
     "avatar": "someOtherUser.png"
   }
}]
```

where "someUser" and "someOtherUser" are people I ("myUserName") subscribed to

Subscriptions:

| from (users)  | to (users)  |
|---|---|
| myUserName  | someUser  |
|  myUserName | someOtherUser  |
| someUser | yetSomeOtherUser |

**Reply by steve-chavez** [reply]
I see, for now you could do this with a SQL function. It should be like:

```sql
create or replace function "suscribedPosts"("myUserName" text) returns setof posts as $$
  select posts.*
  from posts
  join users on users.name = posts.author
  join subscriptions on subscriptions.subscribed = users.name and subscriptions.subscriber = "myUserName" 
$$ language sql;
```

Then you can use this function with `supabase-js` to get your desired JSON shape:

```js
const { data, error } = await supabase
  .rpc('suscribedPosts', {"myUserName": "yourUserName"})
  .select(`
    text,author(name,avatar)
  `);

[{
  "text": "blabla",
  "author": {
     "name": "someUser",
     "avatar": "someUser.png"
   }
},
{
  "text": "hello world",
  "author": {
     "name": "someOtherUser",
     "avatar": "someOtherUser.png"
   }
}]
```
---

Once this [issue](https://github.com/PostgREST/postgrest/issues/1075) is solved, I expect to have this interface that would save the need for a SQL function:

```js
const { data, error } = await supabase
  .from('posts')
  .select(`
    text,author(name,avatar)
  `)
  .eq('users.subscriptions!subscribed.subscriber', 'myUserName') 
```

**Reply by hazae41** [reply]
Thanks 🙏
